### DISCUSSION TITLE
Have been trying for hours to get the simplest shader to work..

### DISCUSSION THREAD

**Discussion by bryancurran** [MonoGame/MonoGame#8397] [Graphics]
I'm about ready to throw ChatGPT into the pool. I've spent at least half a day scouring the internet for any resources on MonoGame shaders/hlsl, even using the WaybackMachine to read tutorials from 2010.  

WHY is everything outside of this light source completely black no matter what I do?  I just want the surroundings to be darker.. like e.g. half the normal brightness.  But no matter what I change, it is completely black.

```
sampler s0;
texture lightMask;  
sampler lightSampler = sampler_state 
{
    Texture = <lightMask>;
    MipFilter = LINEAR;
    MinFilter = LINEAR;
    MagFilter = LINEAR;
};

float brightness = 1.0; // Amount to brighten the RGB channels
float minBrightness = 0.2;

float4 PixelShaderLight(float2 coords : TEXCOORD0) : COLOR0
{
    float4 color = tex2D(s0, coords);
    float4 lightColor = tex2D(lightSampler, coords);

    // Brighten the color by adding the brightness amount to the RGB channels
    float4 brightenedColor = color + float4(brightness, brightness, brightness, 0);

    // Clamp the result to ensure the values are within the valid range [0, 1]
    brightenedColor = clamp(brightenedColor, 0.0, 1.0);

    // Calculate the final color by mixing the brightened color with the light color
    float4 finalColor = brightenedColor * lightColor;

    // Ensure the final color is not completely black by adding a minimum brightness
    finalColor.rgb = max(finalColor.rgb, color.rgb * minBrightness);

    return finalColor;
}

technique Technique1
{
    pass Pass1
    {
        PixelShader = compile ps_2_0 PixelShaderLight();
    }
}
```

![ff6d75866ea33525cea3c5891eb1c7a6](https://github.com/MonoGame/MonoGame/assets/52745033/5927bf46-92bb-40d8-ac8f-8f1232c25360)

In case its relevant, here is my Draw() method from Game1.cs, which has also been refactored a hundred times to try and debug:
```
protected override void Draw(GameTime gameTime)
    {
        GraphicsDevice.SetRenderTarget(mainTarget);
        GraphicsDevice.Clear(Color.CornflowerBlue);

        _spriteBatch.Begin(samplerState: SamplerState.PointClamp, transformMatrix: camera.GetViewMatrix());

        // Draw the level and particles to the main target
        level.Draw(gameTime, _spriteBatch);
        particleSystem.Draw(_spriteBatch);

        _spriteBatch.End();

        // Draw the lights to the light target
        GraphicsDevice.SetRenderTarget(lightTarget);
        GraphicsDevice.Clear(Color.Black);

        _spriteBatch.Begin(samplerState: SamplerState.PointClamp, blendState: BlendState.Additive);
        Vector2 lightPosition = new Vector2(mouseState.X, mouseState.Y);
        float lightScale = 10f;
        _spriteBatch.Draw(lightTexture, lightPosition, null, Color.White, 0f, new Vector2(lightTexture.Width / 2, lightTexture.Height / 2), lightScale, SpriteEffects.None, 0f);

        _spriteBatch.End();

        GraphicsDevice.SetRenderTarget(null);
        GraphicsDevice.Clear(Color.CornflowerBlue);

        // Apply the lighting effect and draw the final image to the back buffer
        lightEffect.Parameters["lightMask"].SetValue(lightTarget);

        _spriteBatch.Begin(effect: lightEffect);
        _spriteBatch.Draw(mainTarget, Vector2.Zero, Color.White);
        _spriteBatch.End();

        base.Draw(gameTime);
    }
```

**Comment by tomspilman** [comment]
My guess is your issue is the input arguments to your pixel shader:

> float4 PixelShaderLight(float2 coords : TEXCOORD0) : COLOR0

Since you're omitting the position and color the texture coordinates likely have unexpected data in it.  

If you make this change i think it will work for you:

> float4 PixelShaderLight(float4 position : SV_POSITION, float2 coords : TEXCOORD0, float4 color : COLOR0) : COLOR0


From the docs (https://docs.monogame.net/articles/content_pipeline/custom_effects.html):

> Make sure the pixel shaders inputs exactly match the vertex shader outputs so the parameters are passed in the correct registers. The parameters need to have the same size and order. Omitting parameters might not break compilation, but can cause unexpected results.

@SimonDarksideJ we likely need to make this more clear with a better tutorial on custom effects.

**Reply by bryancurran** [reply]
Thanks so much for the response.  

So I updated the line you pointed out to use the code you provided.  Initially it caused issues with `mgcb-editor` building it since `float4 color = tex2D(s0, coords);` is already written within the method.  So I commented out that line, and the project now builds and runs.

However, the result is that there is no shadow or light at all.  It's as if the effect doesn't exist (or perhaps the entire screen is just a tad darker, hard to tell):
![6d096094e3fc90fdd767c9f0a73f10d8](https://github.com/MonoGame/MonoGame/assets/52745033/a5938705-3ceb-4733-8123-fb7545f12564)

I don't want to take up too much of your time.  I've read the monogame doc you linked but didn't find it particularly helpful for this. Are there any existing examples of e.g. a Point Light being created in monogame, or somewhere I can read further on that process? Thank you

ETA: Regarding the PixelShader inputs matching the VertexShader outputs, I don't have any VertexShader code within the fx file.  Not sure if relevant, but the entire fx file is pasted in my original post, except it now has the change you suggested.

**Reply by tomspilman** [reply]
> Initially it caused issues with mgcb-editor building it since float4 color = tex2D(s0, coords); is already written within the method.

 I would put that back as you will need it.

 Instead change the input color argument to the pixel shader to be `color_` or something different to avoid the naming conflict.

> However, the result is that there is no shadow or light at all. 

I suspect this is your issue:

```
float brightness = 1.0; // Amount to brighten the RGB channels
float minBrightness = 0.2;
```

... if you change it to:

```
static float brightness = 1.0; // Amount to brighten the RGB channels
static float minBrightness = 0.2;
```

That should fix it.   If you intend to only change these from the shader code and not from C#, then this is the way to go.


> Regarding the PixelShader inputs matching the VertexShader outputs, I don't have any VertexShader code within the fx file

Correct.  But when you write a pixel shader to work with `SpriteBatch` you're using the default vertex shader for `SpriteBatch`.

**Reply by bryancurran** [reply]
Oh my god.. Making the variables `static` fixed it.  Thanks so much.  I would have never in a million years thought that would have an effect.  

`If you intend to only change these from the shader code and not from C#, then this is the way to go.`

Dare I ask, what happens if I do want to change them from C#?

**Reply by tomspilman** [reply]
Don't make them `static`.  And from your C# code do this:

```
lightEffect.Parameters["brightness"].SetValue(1.0f); // Amount to brighten the RGB channels
lightEffect.Parameters["minBrightness"].SetValue(0.2f);
```

**Reply by bryancurran** [reply]
Thanks!

**Comment by SimonDarksideJ** [comment]
It looks like it has been a great learning exercise, and we will surely use this as a test in our updated docs!

Make sure to close the issue @bryancurran if your issue is now resolved, and I hope you join us on reviewing the new docs as soon as they are available.

**Reply by tomspilman** [reply]
FYI.  This wasn't an "issue" it is in Github discussions.

**Reply by SimonDarksideJ** [reply]
![jerkdog](https://github.com/MonoGame/MonoGame/assets/1793042/b7bdc484-08dc-47d2-9f51-9c10a17ec699)

**Comment by stromkos** [comment]
I do hope that the problem is resolved. Just in case someone has a similar problem and lands here.

I just wanted to point out that the multiply should be an addition:

```
 // Ensure the final color is not completely black by adding a minimum brightness
    finalColor.rgb = max(finalColor.rgb, color.rgb * minBrightness);
```

As written: 

If any channel of `color.rgb'  equals `0` no amount multiplied will ever be greater than 0.

`minBrightness` is usually called Ambient Color and applied via addition at the beginning of the calculation.

Your goal is to catch the output and specify a minimum (make sure r=g=b for the `minBrightness` otherwise there will be a non-linear color shift).

```
 // Ensure the final color is not completely black by adding a minimum brightness
    finalColor.rgb = max(finalColor.rgb, color.rgb + minBrightness);
```
